cloud/smartos/vmadm

```diff
@@ -0,0 +1,229 @@
+
+    try:
+        data = json.load(json_data)
+    except:
+        module.fail_json(msg="could not parse json in manifest file: %s" % manifest)
+
+    return data
+
+def strip_keys_from_json(payload):
+    strip_keys = ['last_modified']
+
+    for key in payload.keys():
+        if key in strip_keys:
+            del payload[key]
+
+    return payload
+
+def main():
+    module = AnsibleModule(
+        argument_spec   = dict(
+            state       = dict(choices=['present', 'absent', 'rebooted', 'started', 'stopped', 'updated'], type='str', required=True),
+            uuid        = dict(default=None, aliases=['name'], type='str', required=False),
+            alias       = dict(default=None, type='str', required=False),
+            manifest    = dict(default=None, type='str', required=False),
+            proplist    = dict(default=None, type='str', required=False)
+        ),
+        mutually_exclusive = [['alias', 'uuid'], ['proplist', 'manifest']]
+    )
+
+    # set some useful variables
+    vmadm_path = module.get_bin_path('vmadm', required=True)
+    p = module.params
+    changed = False
+    state, uuid, alias, manifest, proplist = p["state"], p["uuid"], p["alias"], p["manifest"], p["proplist"]
+
+    # vmadm create -f $manifestFile
+    if state == 'present':
+
+        # extra parameter sanity checks (we ignore uuid and alias, as they do nothing here.)
+        check_subparam(module, 'manifest', p, state)
+
+        # abort if we can't find the vm manifest file.
+        if not os.path.isfile(manifest):
+            module.fail_json(msg="could not locate manifest file: %s" % manifest)
+
+        # we need to delve into the manifest and find the alias for the VM, and then
+        # make sure that there is not a VM with that alias (even though SmartOS doesn't
+        # care about duplicated aliases).
+        # Otherwise, a new VM is created each and every single time (not idempotent).
+        data = open_and_load_json_data(module, manifest)
+
+        try:
+            vm_alias = data["alias"]
+        except:
+            module.fail_json(msg="manifest file must provide an alias for the vm")
+
+        if not vmadm_alias_lookup(module, vmadm_path, vm_alias):
+            rc, so, se = module.run_command('%s create -f %s' % (vmadm_path, manifest))
+            changed = True
+        else:
+            rc, so, se = (0, "", "")
+
+    else:
+
+        # -- common checks/tasks for all states except 'present'
+        if not uuid and not alias:
+            module.fail_json(msg="one of 'alias' or 'uuid' required when state is not 'present'")
+
+        if alias:
+            uuid = vmadm_alias_lookup(module, vmadm_path, alias)
+            if not uuid:
+                module.fail_json(msg="could not locate a matching uuid for vm alias '%s'" % alias)
+        # --
+
+        # handles most other states.
+        if state in [ 'started', 'stopped', 'rebooted', 'absent' ]:
```

Comment:
absent does not work at all. absent gets passed to vmadm directly and vmadm dies with invalid command error.